lib/apiservers/engine/backends/convert/stats.go

```diff
@@ -0,0 +1,266 @@
+						cs.config.Cancel()
+					}
+				}
+			}
+		}
+	}()
+	return cs.writer
+}
+
+// ToContainerStats will convert the vic VMMetrics to a docker stats struct -- a complete docker stats
+// struct requires two samples.  Func will return nil until a complete stat is available
+func (cs *ContainerStats) ToContainerStats(current *metrics.VMMetrics) (*types.StatsJSON, error) {
+	// if we have a current metric then validate and transform
+	if cs.currentMetrics != nil {
+		// do we have the same metric as before?
+		if cs.currentMetrics.SampleTime.Equal(current.SampleTime) {
+			// we've already got this as current, so skip and wait for the
+			// next sample
+			return nil, nil
+		}
+		// we have new current stats so need to move the previous CPU
+		err := cs.previousCPU(current)
+		if err != nil {
+			return nil, err
+		}
+	}
+	cs.currentMetrics = current
+
+	// create the current CPU stats
+	cs.currentCPU()
+
+	// create memory stats
+	cs.memory()
+
+	// set sample time
+	cs.dockerStats.Read = cs.currentMetrics.SampleTime
+
+	// PreRead will be zero if we don't have two samples
+	if cs.dockerStats.PreRead.IsZero() {
+		return nil, nil
+	}
+	return cs.dockerStats, nil
+}
+
+func (cs *ContainerStats) memory() {
+	// given KB (i.e. 2048) convert to GB
+	cs.dockerStats.MemoryStats.Limit = uint64(cs.config.Memory * 1000 * 1000)
+	// given KB (i.e. 384.5) convert to Bytes
+	cs.dockerStats.MemoryStats.Usage = uint64(cs.currentMetrics.Memory.Active * 1000)
+}
+
+// previousCPU will move the current stats to the previous CPU location
+func (cs *ContainerStats) previousCPU(current *metrics.VMMetrics) error {
+	// validate that the sampling is in the correct order
+	if current.SampleTime.Before(cs.dockerStats.Read) {
+		err := InvalidOrderError{
+			current:  current.SampleTime,
+			previous: cs.dockerStats.Read,
+		}
+		return err
+	}
+	// move the stats
+	cs.dockerStats.PreCPUStats = cs.dockerStats.CPUStats
+	// set the previousTotal -- this will be added to the current CPU
+	cs.preTotalMhz = cs.dockerStats.PreCPUStats.CPUUsage.TotalUsage
+	cs.dockerStats.PreRead = cs.dockerStats.Read
+	// previous systemUsage will always be the VCH total
+	cs.dockerStats.PreCPUStats.SystemUsage = cs.totalVCHMhz
+
+	return nil
+}
+
+// currentCPU will convert the VM CPU metrics to docker CPU stats
+func (cs *ContainerStats) currentCPU() {
+	cpuCount := len(cs.currentMetrics.CPU.CPUs)
+	dockerCPU := types.CPUStats{
+		CPUUsage: types.CPUUsage{
+			PercpuUsage: make([]uint64, cpuCount, cpuCount),
+		},
+	}
+
+	// collect the current CPU Metrics
+	for ci, current := range cs.currentMetrics.CPU.CPUs {
+		dockerCPU.CPUUsage.PercpuUsage[ci] = uint64(current.MhzUsage)
+		dockerCPU.CPUUsage.TotalUsage += uint64(current.MhzUsage)
+	}
+	// The first stat available for a VM will be missing detail
+	if cpuCount > 0 {
+		dockerCPU.CPUUsage.TotalUsage /= uint64(cpuCount)
+	}
+
+	// Current SystemUsage will always be double the VCH
+	dockerCPU.SystemUsage = cs.dblVCHMhz
```

Comment:
The docker client has a formula that it uses to calculate the CPU usage for a given interval.  So we need to return the data so that the output of the formula makes sense for vSphere. This is something that needs to be documented -- I've started to do that, but don't have it complete.  

If you'd like more detail now let me know.